### PR TITLE
8278173: [vectorapi] Add x64 intrinsics for unsigned (zero extended) casts

### PR DESCRIPTION
**Pull request by merykitty**
Hi,

This patch implements the unsigned upcast intrinsics in x86, which are used in vector lane-wise reinterpreting operations.

Thank you very much.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8278173](https://bugs.openjdk.java.net/browse/JDK-8278173): [vectorapi] Add x64 intrinsics for unsigned (zero extended) casts

### Reviewers
 * [Paul Sandoz](https://openjdk.java.net/census#psandoz) (@PaulSandoz - **Reviewer**)
 * [Sandhya Viswanathan](https://openjdk.java.net/census#sviswanathan) (@sviswa7 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7358/head:pull/7358` \
`$ git checkout pull/7358`

Update a local copy of the PR: \
`$ git checkout pull/7358` \
`$ git pull https://git.openjdk.java.net/jdk pull/7358/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7358`

View PR using the GUI difftool: \
`$ git pr show -t 7358`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7358.diff">https://git.openjdk.java.net/jdk/pull/7358.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/x86/assembler_x86.cpp [hunk 1]
```diff
@@ -4775,6 +4775,15 @@ void Assembler::vpmovzxwd(XMMRegister dst, XMMRegister src, int vector_len) {
   emit_int16(0x33, (0xC0 | encode));
 }
 
+void Assembler::vpmovzxwq(XMMRegister dst, XMMRegister src, int vector_len) {
+  assert(vector_len == AVX_128bit? VM_Version::supports_avx() :
+  vector_len == AVX_256bit? VM_Version::supports_avx2() :
+  vector_len == AVX_512bit? VM_Version::supports_evex() : 0, " ");
+  InstructionAttr attributes(vector_len, /* rex_w */ false, /* legacy_mode */ _legacy_mode_bw, /* no_mask_reg */ true, /* uses_vl */ true);
```
[the remark this answers is not in the document]
Comment (merykitty) [reply src/hotspot/cpu/x86/assembler_x86.cpp]: Fixed, thanks.

File: src/hotspot/cpu/x86/x86.ad [hunk 2]
```diff
@@ -7263,6 +7266,64 @@ instruct vcastDtoL_reg_evex(vec dst, vec src, vec xtmp1, vec xtmp2, kReg ktmp1,
...
+        __ vpmovzxbw($dst$$XMMRegister, $src$$XMMRegister, vlen_enc);
+        break;
+      case T_INT:
+        __ vpmovzxbd($dst$$XMMRegister, $src$$XMMRegister, vlen_enc);
+        break;
+      case T_LONG:
+        __ vpmovzxbq($dst$$XMMRegister, $src$$XMMRegister, vlen_enc);
+        break;
+      default: assert(false, "%s", type2name(to_elem_bt));
+    }
```
[the remark this answers is not in the document]
Comment (merykitty) [reply src/hotspot/cpu/x86/x86.ad]: Fixed, thanks.

**Comment by PaulSandoz** [comment]
Running some tests.

**Comment by PaulSandoz** [comment]
Observing the following failures on CPUs with "Intel_R__Xeon_R__Gold_6354_CPU___3.00GHz" with HotSpot flags:
```
-XX:+CreateCoredumpOnCrash -ea -esa -XX:CompileThreshold=100 -XX:+UnlockExperimentalVMOptions -server -XX:-TieredCompilation
```

TestVectorCastAVX512.java:
```
Failed IR Rules (1)
------------------
- Method "public static void compiler.vectorapi.reshape.tests.TestVectorCast.testUI256toL512(int[],long[])":
  * @IR rule 1: "@compiler.lib.ir_framework.IR(failOn={}, applyIf={}, applyIfAnd={}, applyIfOr={}, counts={"(\\\\d+(\\\\s){2}(VectorUCastI2X.*)+(\\\\s){2}===.*)", "1"}, applyIfNot={})"
    - counts: Graph contains wrong number of nodes:
        Regex 1: (\\d+(\\s){2}(VectorUCastI2X.*)+(\\s){2}===.*)
        Expected 1 but found 0 nodes.
```

TestVectorCastAVX1.java:
```
- Method "public static void compiler.vectorapi.reshape.tests.TestVectorCast.testUB64toS64(byte[],short[])":
  * @IR rule 1: "@compiler.lib.ir_framework.IR(failOn={}, applyIf={}, applyIfAnd={}, applyIfOr={}, counts={"(\\\\d+(\\\\s){2}(VectorUCastB2X.*)+(\\\\s){2}===.*)", "1"}, applyIfNot={})"
    - counts: Graph contains wrong number of nodes:
        Regex 1: (\\d+(\\s){2}(VectorUCastB2X.*)+(\\s){2}===.*)
        Expected 1 but found 0 nodes.

- Method "public static void compiler.vectorapi.reshape.tests.TestVectorCast.testUB64toI128(byte[],int[])":
  * @IR rule 1: "@compiler.lib.ir_framework.IR(failOn={}, applyIf={}, applyIfAnd={}, applyIfOr={}, counts={"(\\\\d+(\\\\s){2}(VectorUCastB2X.*)+(\\\\s){2}===.*)", "1"}, applyIfNot={})"
    - counts: Graph contains wrong number of nodes:
        Regex 1: (\\d+(\\s){2}(VectorUCastB2X.*)+(\\s){2}===.*)
        Expected 1 but found 0 nodes.
```

**Comment by merykitty** [comment]
@PaulSandoz Thanks a lot for your testing, the reason seems to be due to `LaneType::asIntegral` missing `ForceInline` annotation. I have run the reshape test 10 times without getting any failure while with previous patch there is often 1 or 2.
Thanks.

**Comment by PaulSandoz** [comment]
@merykitty testing now passes. Java bits look good. Needs HotSpot reviewer.

**Review by PaulSandoz — APPROVED** [review]

**Review by sviswa7 — APPROVED** [review]

**Comment by sviswa7** [comment]
Hotspot changes look good to me.

**Comment by merykitty** [comment]
Thanks a lot for your reviews.
/integrate

**Pull request closed without merging**